### ISSUE TITLE
asp.net core 2.1 WS Fed proxy

### ISSUE THREAD

**Issue by prophetix** [dotnet/corefx#32346]
Hello,

I am currently trying to move next projects to use .Net Core and we have a corporate proxy using a wpad.dat file for configuration and use the Windows authenticated user for proxy authentication transparently.

Till now, on old .net web application, I used to add default proxy settings in web.config 

```xml
 <system.net>
    <defaultProxy useDefaultCredentials="true">
      <proxy usesystemdefault="False" scriptLocation="http://wpad/wpad.dat" />
    </defaultProxy>
  </system.net>

```
and everything is just fine. 

But I cannot find in .net core 2.1 what i have to add for having the same behavior.

My test project is fairly simple, I created a new .net core 2.1 MVC application. Added the Microsoft.AspNetCore.Authentication.WsFederation nuget package and configured the service :

```c#
        services.AddAuthentication()
    .AddWsFederation(options =>
    {
        options.MetadataAddress = "https://stsserver/2007-06/FederationMetadata.xml";
        options.Wtrealm = "urn:myrealm";
        options.Wreply = "https://localhost:44333/services/login";
    });
```

And when I want to use the new identity provider on the user login screen, I get :

```
An unhandled exception occurred while processing the request.
HttpRequestException: Response status code does not indicate success: 407 (Proxy Authentication Required).
System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()

IOException: IDX20804: Unable to retrieve document from: https://stsserver/2007-06/FederationMetadata.xml'.
Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(string address, CancellationToken cancel)
```

Do you have any Idea on how to do this ?

Thanks.

Thierry

**Comment by karelz** [comment]
@prophetix the config above is for System.Net, yet you use MVC app. Can you please confirm that you hit the same problem in simple console app (no ASP.NET involved)? Can you please post such minimal repro here? (do you use HttpWebRequest or HttpClient)

.NET Core does not have custom WPAD files support AFAIK - either we use system settings (which supports WPAD I think), or you can enter your own proxy info directly in code (no WPAD support).

@wfurt @davidsh should be able to provide more details or correct me ...

**Comment by davidsh** [comment]
Unlike .NET Framework, .NET Core does not support app.config/web.config settings.

The only way to use a PAC (JavaScript) file (either explicit location or discovered via WPAD protocol) is via the IE settings dialog.  And then to pass default network credentials to that proxy, you can use:


```c#
var handler = new HttpClientHandler();
handler.DefaultProxyCredentials = CredentialCache.DefaultCredentials;

var client = new HttpClient(handler);

HttpResponseMessage response = await client.SendAsync(…);  // send request
```

**Comment by prophetix** [comment]
Wpad configuration is already in IE configuration.

I tryed direct connection on Console app using this code ;

```
  static void Main(string[] args)
        {
            TestConnection();
        }

        static private void TestConnection()
        {
            try
            {
                var handler = new HttpClientHandler();
                
                handler.DefaultProxyCredentials = CredentialCache.DefaultCredentials;

                var client = new HttpClient(handler);

                HttpResponseMessage response = client.GetAsync("https://mystsserver.hiddenurl/FederationMetadata/2007-06/FederationMetadata.xml").GetAwaiter().GetResult();
                if (response.IsSuccessStatusCode)
                {
                    var textMessage = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                    Console.WriteLine($"response ok : {textMessage}");
                }
                else
                {
                    Console.WriteLine($"response not ok : {response.ReasonPhrase}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Exception : {ex.Message}");
                throw;
            }
        }
```

- Runing on .Net Core 2.0, no problem 
- Runing on .Net Core 2.1 I get an Exception : ""Authentication failed because the connection could not be reused.""

As it was running on .net core 2.0, I tryed on my MVC app, passing a configured handler to the 


```
            services.AddAuthentication()
    .AddWsFederation(options =>
    {
        options.BackchannelHttpHandler = MyTestHttpHandler;
        options.MetadataAddress = "https://MySTSHiddenURLServer/FederationMetadata/2007-06/FederationMetadata.xml";
        options.Wtrealm = "urn:dotnet:mycustmrmrealm";
        options.Wreply = "https://localhost:44333/services/login";
    });
```

It is working on .Net core 2.0 but fail on .Net core 2.1 with the same error :

HttpRequestException: Authentication failed because the connection could not be reused.
System.Net.Http.HttpConnection.DrainResponseAsync(HttpResponseMessage response)

As i see in the other questions, can it be the same issue with the expect100 issue ?

Do you have a clean way to set the DefaultProxyCredentials property for all new functions calling HttpClientHandler ? 

like the one used in Microsoft.AspNetCore.Authentication.WsFederation nuget package : 

https://github.com/aspnet/Security/blob/master/src/Microsoft.AspNetCore.Authentication.WsFederation/WsFederationPostConfigureOptions.cs

` options.Backchannel = new HttpClient(options.BackchannelHttpHandler ?? new HttpClientHandler());`

**Comment by prophetix** [comment]
Can you tell me where I can found that build / sdk installer ? I do not find it. I get the 2.1.402 but no 2.1.5. And if there is no installer, maybe a list of steps to follow to make it working with VS 2017 ? 

Thanks

**Comment by prophetix** [comment]
> I discussed it with @wfurt. He thinks that the error above "Authentication failed because the connection could not be reused" might be one of the known already addressed issues.
> Can you please try latest (daily) builds of 2.1.5? (it was not publicly released yet) There is a decent chance it has been addressed.
> Let's stick to the simplest repro (the console app without ASP.NET Core). Also, let's avoid wild speculations on expect100 unless you have some evidence that it is related.
> Thanks!

No news about where to find the 2.1.5, have you a direct link to the right download file ? the one on the readme only go to servicing version of 2.1 and some 2.2 or even 3.0, no 2.1.5.

**Comment by caesar1995** [comment]
You can find the steps here: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md to consume a nightly build of .NET Core runtime and SDK.

**Comment by prophetix** [comment]
Just did the test with the 2.1.403 sdk and 2.1.5 runtime, The problem is gone. Good work.  Thanks.

**Comment by KyleLJohnson** [comment]
Is there a place where you can just download 2.1.403 and 2.1.5?

**Comment by davidsh** [comment]
>Is there a place where you can just download 2.1.403 and 2.1.5? I need to install on my build server.

We currently don't have a convenient place for you to download the 2.1.5 servicing release since it has not been released yet.  We don't publish daily builds of 2.1.x.

But all fixes that go into 2.1.x also get put into the pending 2.2 release.  And you can get daily builds of that here:

https://github.com/dotnet/core-sdk/blob/master/README.md#installers-and-binaries
https://dotnetcli.blob.core.windows.net/dotnet/Sdk/release/2.2.1xx/dotnet-sdk-latest-win-x64.exe

You'll need to re-target your application to 'netcore2.2' to take advantage of that.

**Comment by davidsh** [comment]
>Just did the test with the 2.1.403 sdk and 2.1.5 runtime, The problem is gone. Good work. Thanks.

Glad to hear that the problem is fixed.  I'll close this issue now.  Thanks for verifying!
